massiv/src/Data/Massiv/Core/Index/Class.hs

```diff
@@ -19,14 +21,21 @@ module Data.Massiv.Core.Index.Class where
 import           Control.DeepSeq           (NFData (..))
 import           Data.Functor.Identity     (runIdentity)
 import           Data.Massiv.Core.Iterator
+import           Data.Validity
 import           GHC.TypeLits
+import           GHC.Generics (Generic)
 
 -- | A way to select Array dimension.
 newtype Dim = Dim Int deriving (Show, Eq, Ord, Num, Real, Integral, Enum)
 
+instance Validity Dim where
+  validate (Dim n) = declare "A Dim is positive" $ n >= 0
```

Comment:
`Dim` indeed is positive, since we don't deal with zero dimensional arrays, so shouldn't it be `n > 0`? At least, as far as I remember, there is no place in the package, where n = 0 is considered valid.

Reply:
@lehins Why not store a `Word` or `Natural` then? (optionally with the semantics that `Dim 0` means `1`?

Reply:
Using `Word` doesn't protect you from negative values, it simply wraps around, which is terrible in my opinion. Attempt to construct a negative `Natural` results in a runtime underflow exception, which is not much better. In every place `Dim` is used, an invalid value will result in a sensible error message. Not only negative values will result in an error, but in case `Dim` being higher than dimensionality of the array will as well. Plus there are functions that will return `Nothing` in above cases, which would become impossible for `Natural` in pure setting.

Reply:
By the way, this reminds me that there could be instances as well: `Index ix => RelativeValidity Dim (Array r ix e)` and `Index ix => RelativeValidity Dim ix`

Reply:
That's precisely the point. `Ix0` does NOT have an instance of `Index` class and is not considered a valid index by the library. It is only needed only for the type system to be happy, since `Lower Ix1 = Ix0`. For the same reason, there is no place in the library where `Dim 0` would be valid either.